### PR TITLE
Add groups

### PR DESCRIPTION
**Pull request by ShirleyWangxt**

### CODE REVIEW

File: controllers/groups_controller.js [hunk 1]
```diff
@@ -0,0 +1,101 @@
+const db = require('../db');
+const sql = require('../sql');
+
+exports.index = (req, res) => {
+  db.query(sql.groups.queries.get_groups, (err, data) => {
+    if (err) console.error('Cannot get groups');
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]: Handle the error by calling `next(err)`

File: controllers/groups_controller.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    res.render('groups', { data: data.rows });
+  });
+};
+
+exports.new = (req, res) => {
+  db.query(sql.courses.queries.get_courses, (err, courses) => {
+    if (err) console.error('Cannot get courses');
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]: same here

File: controllers/groups_controller.js [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    res.render('groupNew', { courses: courses.rows });
+  });
+};
+
+exports.create = (req, res) => {
+  const { name } = req.body;
+  const { semester_name } = req.body;
+  const { module_code } = req.body;
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]: Can be simplified the three lines to this I think?
```suggestion
  const { name, semester_name, module_code } = req.body;
```

File: controllers/groups_controller.js [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  const { name } = req.body;
+  const { semester_name } = req.params;
+  const { module_code } = req.params;
+  db.query(sql.groups.queries.find_group, [name, semester_name, module_code], (err, data) => {
+    if (err) {
+      console.error('Cannot find group');
+      next(err);
+    } else {
+      // TODO: refactor
+      const group = {
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]:
```suggestion
      const group = data.rows[0];
```

File: controllers/groups_controller.js (hunk 1 above)
Comment (Na-Nazhou) [reply controllers/groups_controller.js]: Change `console.error` to `log.error` according to #55

File: controllers/groups_controller.js [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  const { module_code } = req.body;
+
+  db.query(
+    sql.groups.queries.update_group,
+    [name, semester_name, module_code, old_semester_name, old_module_code],
+    err => {
+      if (err) {
+        console.error('Failed to update group');
+        req.flash('error', err.message);
+        res.redirect(`/${encodeURIComponent(semester_name)}/${encodeURIComponent(module_code)}/edit`);
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]: Is the redirect URL correct?

File: routes/index.js [hunk 6]
```diff
@@ -24,6 +25,7 @@ router.use('/users', ensureAuthenticated, users);
 router.use('/semesters', ensureAuthenticated, ensureIsAdmin, semesters);
 router.use('/modules', ensureAuthenticated, ensureIsAdmin, modules);
 router.use('/courses', ensureAuthenticated, courses);
+router.use('/groups', ensureAuthenticated, ensureProfessor, groups);
```
Comment (gary-lgy) [inline routes/index.js]: The routes for groups should be nested within course. See #51 as an example.

**Review by gary-lgy — CHANGES_REQUESTED** [review]

File: controllers/groups_controller.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+};
+
+exports.new = (req, res) => {
+  const { semester_name, module_code } = req.params;
+  res.render('groupForm', { semester_name, module_code, group: null });
+};
+
+exports.create = (req, res) => {
+  const { semester_name, module_code, name } = req.body;
+  log.fatal('', { semester_name, module_code, name });
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]:
```suggestion
```

File: controllers/groups_controller.js [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+  });
+};
+
+exports.update = (req, res) => {
+  const { semester_name, module_code, name: old_name } = req.params;
+  const { name: new_name } = req.body;
+
+  db.query(sql.groups.queries.update_group, [semester_name, module_code, old_name, new_name], err => {
+    if (err) {
+      log.error('Failed to update course');
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]:
```suggestion
      log.error('Failed to update group');
```

File: routes/groups.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+const groups = require('../controllers/groups_controller');
+const log = require('../helpers/logging');
+
+router.use((req, res, next) => {
+  log.controller('groups controller handling the request');
+  next();
+});
+
+router.get('/new', groups.new);
+router.get('/:name', groups.show);
```
Comment (Na-Nazhou) [inline routes/groups.js:11]: Do we need the group#show page? 

Edit: ok nvm i think we will put a list of users in this group in the group#show page

File: routes/index.js [hunk 10]
```diff
@@ -4,7 +4,8 @@ const users = require('./users');
 const semesters = require('./semesters');
 const modules = require('./modules');
 const courses = require('./courses');
-const { ensureAuthenticated, ensureIsAdmin } = require('../auth/middleware');
+
+const { ensureAuthenticated, ensureIsAdmin, ensureProfessor } = require('../auth/middleware');
```
Comment (Na-Nazhou) [inline routes/index.js]: Can you revert the changes here? I am going to refactor the permissions

**Review by Na-Nazhou — CHANGES_REQUESTED** [review]

File: controllers/groups_controller.js [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+exports.new = (req, res) => {
+  const { semester_name, module_code } = req.params;
+  res.render('groupForm', { semester_name, module_code, group: null });
+};
+
+exports.create = (req, res) => {
+  const { semester_name, module_code, name } = req.body;
+  db.query(sql.groups.queries.create_group, [semester_name, module_code, name], err => {
+    if (err) {
+      log.error('Failed to create groups');
```
Comment (Na-Nazhou) [inline controllers/groups_controller.js]:
```suggestion
      log.error('Failed to create group');
```

**Review by Na-Nazhou — CHANGES_REQUESTED** [review]
Just a minor fix

File: routes/index.js [hunk 12]
```diff
@@ -4,6 +4,7 @@ const users = require('./users');
 const semesters = require('./semesters');
 const modules = require('./modules');
 const courses = require('./courses');
+
```
Comment (Na-Nazhou) [inline routes/index.js]:
```suggestion
```

**Review by Na-Nazhou — APPROVED** [review]

**Pull request merged**